mmcls/models/backbones/levit.py

```diff
@@ -0,0 +1,545 @@
+        x = self.linear1(x)
+        x = self.activation(x)
+        x = self.linear2(x)
+        return x
+
+
+class Subsample(BaseModule):
+
+    def __init__(self, stride, resolution):
+        super(Subsample, self).__init__()
+        self.stride = stride
+        self.resolution = resolution
+
+    def forward(self, x):
+        B, N, C = x.shape  # 2 196 128
+        x = x.view(B, self.resolution, self.resolution,
+                   C)[  # 2 196 128 -> 2 14 14 128  -> 2 7 7 128
+                       :, ::self.stride, ::self.stride].reshape(
+                           B, -1, C)  # # 2 14 14 128 > 2 49 128
+        return x
+
+
+class AttentionSubsample(nn.Sequential):
+
+    def __init__(self,
+                 in_dim,
+                 out_dim,
+                 key_dim,
+                 num_heads=8,
+                 attn_ratio=2,
+                 activation=None,
+                 stride=2,
+                 resolution=14,
+                 resolution_=7):
+        super(AttentionSubsample, self).__init__()
+        self.num_heads = num_heads
+        self.scale = key_dim**-0.5
+        self.key_dim = key_dim
+        self.nh_kd = nh_kd = key_dim * num_heads
+        self.d = int(attn_ratio * key_dim)
+        self.dh = int(attn_ratio * key_dim) * self.num_heads
+        self.attn_ratio = attn_ratio
+        self.resolution_ = resolution_
+        self.resolution_2 = resolution_**2
+        h = self.dh + nh_kd
+        self.kv = LinearBatchNorm(in_dim, h)
+
+        self.q = nn.Sequential(
+            Subsample(stride, resolution), LinearBatchNorm(in_dim, nh_kd))
+        self.proj = nn.Sequential(activation(),
+                                  LinearBatchNorm(self.dh, out_dim))
+
+        self.stride = stride
+        self.resolution = resolution
+        points = list(itertools.product(range(resolution), range(resolution)))
+        points_ = list(
+            itertools.product(range(resolution_), range(resolution_)))
+        N = len(points)
+        N_ = len(points_)
+        attention_offsets = {}
+        idxs = []
+        for p1 in points_:
+            for p2 in points:
+                size = 1
+                offset = (abs(p1[0] * stride - p2[0] + (size - 1) / 2),
+                          abs(p1[1] * stride - p2[1] + (size - 1) / 2))
+                if offset not in attention_offsets:
+                    attention_offsets[offset] = len(attention_offsets)
+                idxs.append(attention_offsets[offset])
+        self.attention_biases = torch.nn.Parameter(
+            torch.zeros(num_heads, len(attention_offsets)))
+        self.register_buffer('attention_bias_idxs',
+                             torch.LongTensor(idxs).view(N_, N))
+
+    @torch.no_grad()
+    def train(self, mode=True):
+        super(AttentionSubsample, self).train(mode)
+        if mode and hasattr(self, 'ab'):
+            del self.ab
+        else:
+            self.ab = self.attention_biases[:, self.attention_bias_idxs]
```

Comment:
What's it for?

Reply:
I understand that the trained attention_bias is used during testing, and the structure needs to be deleted during training. This structure is also used in the efficientformer